File: tenants/yeppo/scripts/extract_crisp.js
/**
 * extract_crisp.js — Extrae conversaciones de Crisp y genera knowledge base
 *
 * Uso: node extract_crisp.js
 */

const axios = require('axios');
const fs    = require('fs');
const path  = require('path');

const WEBSITE_ID = process.env.CRISP_WEBSITE_ID || '0aaff274-e16d-4985-8d59-76e715e12e72';
const IDENTIFIER = process.env.CRISP_IDENTIFIER || 'b37a978c-dc73-47f7-8be7-6dda4dd8f2bd';
const API_KEY    = process.env.CRISP_API_KEY    || '388f2e24cd2ab3d399339850759c40c3245c85017367807dbd3b05de8fab7cc8';

const AUTH = Buffer.from(`${IDENTIFIER}:${API_KEY}`).toString('base64');

const client = axios.create({
  baseURL: 'https://api.crisp.chat/v1',
  headers: { Authorization: `Basic ${AUTH}`, 'X-Crisp-Tier': 'plugin' },
  timeout: 15000
});

const OUT_DIR = path.join(__dirname, '..', 'knowledge');
if (!fs.existsSync(OUT_DIR)) fs.mkdirSync(OUT_DIR, { recursive: true });

const DELAY_MS = 250;
const MAX_PAGES = 30;

function sleep(ms) {
  return new Promise(r => setTimeout(r, ms));
}

function cleanText(text) {
  if (!text || typeof text !== 'string') return '';
  return text.replace(/https?:\/\/\S+/g, '[link]').replace(/\n{3,}/g, '\n\n').trim();
}

async function run() {
  console.log('📥 Descargando conversaciones...');

  // 1. Listar todas las conversaciones
  const allConvs = [];
  for (let page = 1; page <= MAX_PAGES; page++) {
    try {
      const res = await client.get(`/website/${WEBSITE_ID}/conversations/${page}`);
      const data = res.data.data;
      if (!data || data.length === 0) break;
      allConvs.push(...data);
      console.log(`  Página ${page}: ${data.length} (total: ${allConvs.length})`);
      if (data.length < 20) break;
      await sleep(DELAY_MS);
    } catch (e) {
      console.error(`  Error página ${page}:`, e.message);
      break;
    }
  }
  console.log(`✅ ${allConvs.length} conversaciones encontradas\n`);

  // 2. Descargar mensajes de cada una
  console.log('📨 Descargando mensajes...');
  const knowledge = [];
  let skipped = 0;

  for (let i = 0; i < allConvs.length; i++) {
    const conv = allConvs[i];
    const sid  = conv.session_id;

    if (i % 50 === 0) console.log(`  [${i}/${allConvs.length}] procesadas...`);

    try {
      const res  = await client.get(`/website/${WEBSITE_ID}/conversation/${sid}/messages`);
      const msgs = res.data.data || [];

      // Solo mensajes de texto con contenido
      const textMsgs = msgs.filter(m =>
        m.type === 'text' &&
        m.content &&
        typeof m.content === 'string' &&
        m.content.trim().length > 2 &&
        m.content.trim().length < 1500
      );

      // Requiere al menos 1 mensaje del operador
      if (!textMsgs.some(m => m.from === 'operator')) {
        skipped++;
        await sleep(DELAY_MS);
        continue;
      }

      const dialogue = textMsgs
        .map(m => `${m.from === 'operator' ? 'Agente' : 'Cliente'}: ${cleanText(m.content)}`)
        .join('\n');

      knowledge.push({
        channel:    conv.meta && conv.meta.origin ? conv.meta.origin : 'unknown',
        date:       new Date(conv.created_at).toISOString().split('T')[0],
        dialogue
      });

    } catch (e) {
      skipped++;
    }

    await sleep(DELAY_MS);
  }

  console.log(`\n✅ Con respuesta del equipo: ${knowledge.length}`);
  console.log(`⏭️  Sin respuesta (omitidas): ${skipped}`);

  // 3. Guardar
  const kbPath = path.join(OUT_DIR, 'knowledge_base.json');
  fs.writeFileSync(kbPath, JSON.stringify(knowledge, null, 2), 'utf8');
  console.log(`\n💾 Guardado: ${kbPath}`);

  // 4. Stats por canal
  const channels = {};
  knowledge.forEach(c => { channels[c.channel] = (channels[c.channel] || 0) + 1; });
  console.log('\n📊 Por canal:');
  Object.entries(channels).sort((a, b) => b[1] - a[1]).forEach(([ch, n]) => {
    console.log(`   ${ch}: ${n}`);
  });

  console.log('\n🎉 Listo!');
}

run().catch(e => {
  console.error('❌ Error fatal:', e.message);
  process.exit(1);
});

File: tenants/yeppo/scripts/extract_crisp.py
"""
extract_crisp.py — Extrae conversaciones de Crisp y genera knowledge base
Uso: python extract_crisp.py
"""

import urllib.request
import urllib.error
import json
import base64
import time
import os
import sys

WEBSITE_ID = "0aaff274-e16d-4985-8d59-76e715e12e72"
IDENTIFIER = "b37a978c-dc73-47f7-8be7-6dda4dd8f2bd"
API_KEY    = "388f2e24cd2ab3d399339850759c40c3245c85017367807dbd3b05de8fab7cc8"

AUTH = base64.b64encode(f"{IDENTIFIER}:{API_KEY}".encode()).decode()
HEADERS = {"Authorization": f"Basic {AUTH}", "X-Crisp-Tier": "plugin"}

DELAY    = 0.25
MAX_PAGES = 30

OUT_DIR = os.path.join(os.path.dirname(__file__), "..", "knowledge")
os.makedirs(OUT_DIR, exist_ok=True)

def api_get(endpoint):
    url = f"https://api.crisp.chat/v1{endpoint}"
    req = urllib.request.Request(url, headers=HEADERS)
    with urllib.request.urlopen(req, timeout=15) as r:
        return json.loads(r.read().decode())["data"]

def clean(text):
    if not text or not isinstance(text, str):
        return ""
    import re
    text = re.sub(r"https?://\S+", "[link]", text)
    text = re.sub(r"\n{3,}", "\n\n", text)
    return text.strip()

# 1. Listar conversaciones
print("📥 Descargando conversaciones...")
all_convs = []
for page in range(1, MAX_PAGES + 1):
    try:
        data = api_get(f"/website/{WEBSITE_ID}/conversations/{page}")
        if not data:
            break
        all_convs.extend(data)
        print(f"  Página {page}: {len(data)} (total: {len(all_convs)})", flush=True)
        if len(data) < 20:
            break
        time.sleep(DELAY)
    except Exception as e:
        print(f"  Error página {page}: {e}")
        break

print(f"✅ {len(all_convs)} conversaciones\n")

# 2. Mensajes
print("📨 Descargando mensajes...")
knowledge = []
skipped   = 0

for i, conv in enumerate(all_convs):
    sid = conv["session_id"]

    if i % 50 == 0:
        print(f"  [{i}/{len(all_convs)}] procesadas...", flush=True)

    try:
        msgs = api_get(f"/website/{WEBSITE_ID}/conversation/{sid}/messages")
        if not msgs:
            skipped += 1
            time.sleep(DELAY)
            continue

        text_msgs = [
            m for m in msgs
            if m.get("type") == "text"
            and isinstance(m.get("content"), str)
            and 2 < len(m["content"].strip()) < 1500
        ]

        if not any(m.get("from") == "operator" for m in text_msgs):
            skipped += 1
            time.sleep(DELAY)
            continue

        dialogue = "\n".join(
            f"{'Agente' if m['from'] == 'operator' else 'Cliente'}: {clean(m['content'])}"
            for m in text_msgs
        )

        origin = ""
        if conv.get("meta") and conv["meta"].get("origin"):
            origin = conv["meta"]["origin"]

        from datetime import datetime, timezone
        ts = conv["created_at"] / 1000
        date = datetime.fromtimestamp(ts, tz=timezone.utc).strftime("%Y-%m-%d")

        knowledge.append({"channel": origin or "unknown", "date": date, "dialogue": dialogue})

    except Exception as e:
        skipped += 1

    time.sleep(DELAY)

print(f"\n✅ Con respuesta del equipo: {len(knowledge)}")
print(f"⏭️  Omitidas: {skipped}")

# 3. Guardar
kb_path = os.path.join(OUT_DIR, "knowledge_base.json")
with open(kb_path, "w", encoding="utf-8") as f:
    json.dump(knowledge, f, ensure_ascii=False, indent=2)
print(f"\n💾 Guardado: {kb_path}")

# 4. Stats
from collections import Counter
channels = Counter(c["channel"] for c in knowledge)
print("\n📊 Por canal:")
for ch, n in channels.most_common():
    print(f"   {ch}: {n}")

print("\n🎉 Listo!")
